tsdb/index/labelvalues.go

```diff
@@ -0,0 +1,332 @@
+
+func (r *Reader) labelValuesForV1(postings Postings, name string, inverted bool) storage.LabelValues {
+	e := r.postingsV1[name]
+	if len(e) == 0 {
+		return storage.EmptyLabelValues()
+	}
+	vals := make([]string, 0, len(e))
+	for v := range e {
+		vals = append(vals, v)
+	}
+	slices.Sort(vals)
+	return &intersectLabelValuesV1{
+		seriesRefs: e,
+		values:     vals,
+		postings:   NewPostingsCloner(postings),
+		b:          r.b,
+		dec:        r.dec,
+		inverted:   inverted,
+	}
+}
+
+type intersectLabelValuesV1 struct {
+	seriesRefs map[string]uint64
+	values     []string
+	postings   *PostingsCloner
+	b          ByteSlice
+	dec        *Decoder
+	cur        string
+	err        error
+	inverted   bool
+}
+
+func (it *intersectLabelValuesV1) Next() bool {
+	if it.err != nil {
+		return false
+	}
+
+	// Look for a value with intersecting postings
+	for len(it.values) > 0 {
+		val := it.values[0]
+		it.values = it.values[1:]
+
+		postingsOff := it.seriesRefs[val]
+		// Read from the postings table.
+		d := encoding.NewDecbufAt(it.b, int(postingsOff), castagnoliTable)
+		_, curPostings, err := it.dec.Postings(d.Get())
+		if err != nil {
+			it.err = fmt.Errorf("decode postings: %w", err)
+			return false
+		}
+
+		isMatch := intersect(curPostings, it.postings.Clone())
+		if it.inverted {
+			isMatch = !isMatch
+		}
+		if isMatch {
+			it.cur = val
+			return true
+		}
+	}
+
+	return false
+}
+
+func (it *intersectLabelValuesV1) At() string {
+	return it.cur
+}
+
+func (it *intersectLabelValuesV1) Err() error {
+	return it.err
+}
+
+func (it *intersectLabelValuesV1) Warnings() annotations.Annotations {
+	return nil
+}
+
+func (it *intersectLabelValuesV1) Close() error {
+	return nil
+}
+
+type intersectLabelValues struct {
+	d         *encoding.Decbuf
+	b         ByteSlice
+	dec       *Decoder
+	postings  *PostingsCloner
+	lastVal   string
+	skip      int
+	cur       string
+	exhausted bool
+	err       error
+	inverted  bool
+}
+
+func (it *intersectLabelValues) Next() bool {
+	if it.exhausted || it.err != nil {
+		return false
+	}
+
+	for !it.exhausted && it.d.Err() == nil {
+		if it.skip == 0 {
+			// These are always the same number of bytes,
+			// and it's faster to skip than to parse.
+			it.skip = it.d.Len()
+			// Key count
+			it.d.Uvarint()
+			// Label name
+			it.d.UvarintBytes()
+			it.skip -= it.d.Len()
+		} else {
+			it.d.Skip(it.skip)
+		}
+
+		// Label value
+		vb := it.d.UvarintBytes()
+		v := yoloString(vb)
+
+		postingsOff := int(it.d.Uvarint64())
+		// Read from the postings table
+		postingsDec := encoding.NewDecbufAt(it.b, postingsOff, castagnoliTable)
+		_, curPostings, err := it.dec.Postings(postingsDec.Get())
+		if err != nil {
+			it.err = fmt.Errorf("decode postings: %w", err)
+			return false
+		}
+
+		it.exhausted = v == it.lastVal
+
+		isMatch := intersect(curPostings, it.postings.Clone())
+		if it.inverted {
+			isMatch = !isMatch
+		}
+		if isMatch {
+			// Make sure to allocate a new string
+			it.cur = string(vb)
```

Comment:
There's no need for v/vb, because `string(val) == it.lastVal` doesn't allocate.
So we can do just this: (I called the variable `val` as opposed to `lastVal`)

```suggestion
		val := it.d.UvarintBytes()
		postingsOff := int(it.d.Uvarint64())
		// Read from the postings table
		postingsDec := encoding.NewDecbufAt(it.b, postingsOff, castagnoliTable)
		_, curPostings, err := it.dec.Postings(postingsDec.Get())
		if err != nil {
			it.err = fmt.Errorf("decode postings: %w", err)
			return false
		}

		it.exhausted = string(val) == it.lastVal

		isMatch := intersect(curPostings, it.postings.Clone())
		if it.inverted {
			isMatch = !isMatch
		}
		if isMatch {
			// Make sure to allocate a new string
			it.cur = string(val)
```

Reply:
@colega for my understanding - `string(val) == it.lastVal` doesn't allocate unlike `s := string(val)`? Otherwise, why would you need `yoloString`, right?

Reply:
I applied your suggestion manually btw. Can't commit the suggestion directly due to changes in the meantime.

Reply:
`string(val) == it.lastVal` doesn't allocate because the compiler optimizes that into just comparing the underlying values.